[GStreamer][WebRTC] Rewrite stats timestamps

philn opened the pull request:
#### fa7d13583e7fecaeef8209f47ca396e7e026e71d
<pre>
[GStreamer][WebRTC] Rewrite stats timestamps
<a href="https://bugs.webkit.org/show_bug.cgi?id=279508">https://bugs.webkit.org/show_bug.cgi?id=279508</a>

Reviewed by Xabier Rodriguez-Calvar.

The stats provided by webrtcbin are timestamped using the system monotonic time, which is not
exactly following the spec which stipulates that we should use Performance.timeOrigin +
Performance.now(). By applying the monotonic offset to the epoch we can use
Performance::reduceTimeResolution() and get a bit closer to spec compliance.

Covered by a new test, inspired from
imported/w3c/web-platform-tests/webrtc/RTCPeerConnection-getStats.https.html that checks the
timestamp remain close to window.performance.timeOrigin + window.performance.now().

* LayoutTests/webrtc/stats-timestamp-increases-expected.txt: Added.
* LayoutTests/webrtc/stats-timestamp-increases.html: Added.
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp:
(WebCore::GStreamerMediaEndpoint::preprocessStats):
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerStatsCollector.cpp:
(WebCore::RTCStatsReport::Stats::Stats):
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerWebRTCUtils.cpp:
(WebCore::StatsTimestampConverter::singleton):
(WebCore::StatsTimestampConverter::convertFromMonotonicTime const):
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerWebRTCUtils.h:

Canonical link: <a href="https://commits.webkit.org/283693@main">https://commits.webkit.org/283693@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ff43da22ef77d8e6c5a79bdec88a3c4a5bed9d0e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/66621 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/45996 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/19243 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/70655 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/17754 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/68739 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/53795 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/17517 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/53386 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/11977 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/69688 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/42348 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/57652 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/34053 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/39019 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/15043 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/16108 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/60930 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/15384 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/72357 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/10578 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/14746 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/60714 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/10610 "Built successfully") | [❌ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/57723 "Exiting early after 10 failures. 100 tests run. 1 flakes") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/61045 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14820 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/8706 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/2326 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/41803 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/42880 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/44063 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/42623 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cadubentzen reviewed (commented):
Would the timestamps of two different PeerConnection objects be comparable like this? The `Performance.timeOrigin` bit would come from `StatsTimestampConverter::m_epoch` that is created with every new GStreamerMediaEndpoint from the PC object. I'm wondering whether `Performance.timeOrigin`  should be the same between objects. From https://www.w3.org/TR/hr-time-3/#dfn-get-time-origin-timestamp:

> To get time origin timestamp, given a [global object](https://html.spec.whatwg.org/multipage/webappapis.html#global-object) global, run the following steps, which return a [duration](https://www.w3.org/TR/hr-time-3/#dfn-duration):
> 
> 1.  Let timeOrigin be global's [relevant settings object](https://html.spec.whatwg.org/multipage/webappapis.html#relevant-settings-object)'s [time origin](https://html.spec.whatwg.org/multipage/webappapis.html#concept-settings-object-time-origin).
>
> Note
> In [Window](https://html.spec.whatwg.org/multipage/nav-history-apis.html#window) contexts, this value represents the time when [navigation has started](https://html.spec.whatwg.org/multipage/browsing-the-web.html#navigate). In [Worker](https://html.spec.whatwg.org/multipage/workers.html#worker) and [ServiceWorker](https://www.w3.org/TR/service-workers/#serviceworker) contents, this value represent the time when the [worker is run](https://html.spec.whatwg.org/multipage/workers.html#run-a-worker). [[service-workers](https://www.w3.org/TR/hr-time-3/#bib-service-workers)]

cadubentzen reviewed (approved):
Nice! (not official r+, not a reviewer)

cadubentzen commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp:
```diff
@@ -1893,6 +1893,13 @@ GUniquePtr<GstStructure> GStreamerMediaEndpoint::preprocessStats(const GRefPtr<G
+        auto timestamp = gstStructureGet<double>(structure.get(), "timestamp"_s);
+        if (LIKELY(timestamp)) {
+            auto newTimestamp = StatsTimestampConverter::singleton().convertFromMonotonicTime(Seconds::fromMilliseconds(*timestamp));
+            gst_structure_set(structure.get(), "timestamp", G_TYPE_DOUBLE, newTimestamp.microseconds(), nullptr);
```
?
```suggestion
            gst_structure_set(structure.get(), "timestamp", G_TYPE_DOUBLE, newTimestamp.milliseconds(), nullptr);
```

cadubentzen replied:
I'm a bit confused, aren't the timestamps in milliseconds from webrtcbin's `get-stats`?

philn replied:
Yes they are indeed. We convert to microseconds here to be consistent with libwebrtc, which iiuc internally uses microseconds too, hence the JSON logs are timestamped using the same semantics for both backends.

cadubentzen replied:
Thanks for the clarification!

philn committed:
[GStreamer][WebRTC] Rewrite stats timestamps
https://bugs.webkit.org/show_bug.cgi?id=279508

Reviewed by Xabier Rodriguez-Calvar.

The stats provided by webrtcbin are timestamped using the system monotonic time, which is not
exactly following the spec which stipulates that we should use Performance.timeOrigin +
Performance.now(). By applying the monotonic offset to the epoch we can use
Performance::reduceTimeResolution() and get a bit closer to spec compliance.

Covered by a new test, inspired from
imported/w3c/web-platform-tests/webrtc/RTCPeerConnection-getStats.https.html that checks the
timestamp remain close to window.performance.timeOrigin + window.performance.now().

* LayoutTests/webrtc/stats-timestamp-increases-expected.txt: Added.
* LayoutTests/webrtc/stats-timestamp-increases.html: Added.
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp:
(WebCore::GStreamerMediaEndpoint::preprocessStats):
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerStatsCollector.cpp:
(WebCore::RTCStatsReport::Stats::Stats):
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerWebRTCUtils.cpp:
(WebCore::StatsTimestampConverter::singleton):
(WebCore::StatsTimestampConverter::convertFromMonotonicTime const):
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerWebRTCUtils.h:

Canonical link: https://commits.webkit.org/283693@main
